types/src/ledger_info.rs

```diff
@@ -373,6 +374,192 @@ impl LedgerInfoWithPartialSignatures {
+    ) {
+        if self
+            .verified_signatures
+            .signatures()
+            .contains_key(&validator)
+        {
+            return;
+        }
+        let entry = self.unverified_signatures.entry(validator).or_default();
+        if !entry.contains(&signature) {
+            entry.push(signature);
+        }
+    }
+
+    pub fn num_unverified_signatures(&self) -> usize {
+        self.unverified_signatures.len()
+    }
+
+    pub fn verified_voters(&self) -> Vec<&AccountAddress> {
+        self.verified_signatures
+            .signatures()
+            .keys()
+            .collect_vec()
+            .clone()
+    }
+
+    pub fn unverified_voters(&self) -> Vec<&AccountAddress> {
+        self.unverified_signatures.keys().collect()
+    }
+
+    pub fn aggregated_voters(&self) -> usize {
+        self.verified_aggregate_signature.get_num_voters()
+    }
+
+    // Collecting all the authors from verified signatures, unverified signatures and the aggregated signature.
+    pub fn all_voters(&self, verifier: &ValidatorVerifier) -> Vec<AccountAddress> {
+        let authors = self.verified_aggregate_signature.get_signers_addresses(
+            verifier
+                .get_ordered_account_addresses_iter()
+                .collect_vec()
+                .as_slice(),
+        );
+        let authors = authors
+            .iter()
+            .chain(self.verified_signatures.signatures().keys());
+        let authors = authors
+            .chain(self.unverified_signatures.keys())
+            .collect::<HashSet<_>>();
+        authors.iter().cloned().cloned().collect()
+    }
+
+    pub fn check_voting_power(
+        &self,
+        verifier: &ValidatorVerifier,
+    ) -> std::result::Result<u128, VerifyError> {
+        let all_voters = self.all_voters(verifier);
+        verifier.check_voting_power(all_voters.iter().collect_vec().into_iter(), true)
+    }
+
+    pub fn aggregate_and_verify(
+        &mut self,
+        verifier: &ValidatorVerifier,
+    ) -> Result<LedgerInfoWithSignatures, VerifyError> {
+        self.check_voting_power(verifier)?;
+
+        // Compute signatures from verified signatures and unverified signatures that are not already
+        // included in the verified aggregate signature.
+        let mut partial_signatures = self.verified_signatures.clone();
+        for (author, signatures) in self.unverified_signatures.iter() {
+            partial_signatures.add_signature(*author, signatures[0].clone());
+        }
+        let authors_in_aggregated_sig = self.verified_aggregate_signature.get_signers_addresses(
+            verifier
+                .get_ordered_account_addresses_iter()
+                .collect_vec()
+                .as_slice(),
+        );
+        for author in authors_in_aggregated_sig {
+            partial_signatures.remove_signature(author);
+        }
+
+        if partial_signatures.is_empty() {
+            return Ok(LedgerInfoWithSignatures::new(
```

Comment:
this doesn't make much sense, if all signatures come in verified, this returns Okay without doing aggregations, the verified_aggregate_signature is empty afaiu

Reply:
This case shouldn't ideally occur.
Basically it's checking that if the aggreatgate voting power exceeds the treshold, and if no new signatures are available in verified_signatures and unverified_signatures, it means the verified_aggregate_signature is already good enough. So, it just responds with verified_aggregate_signature.

Reply:
both partial_signatures and verified_aggregate_signature can be empty while verified_signatures have enough voting powers, this just return an empty signature. (in the case that we verify all signatures individually)

Reply:
`let mut partial_signatures = self.verified_signatures.clone();`
This line makes sure partial_signatures contain all the verified_signatures. In verified_signatures have all the votes, partial_signatures won't be empty.